### ISSUE TITLE
Negative number of reads?

### ISSUE THREAD

**Issue by StefanoLonardi** [broadinstitute/pilon#15]
I am using PILON To correct a 500M genome assembly obtained from CANU+Quiver. I have 
2,190,431,976 reads in sorted.bam file (should guarantee about 300x coverage). When I start PILON, it reports a negative number of reads (see below). Also the coverage reported for each unitig does not seem right (I am expecting ~300x).

```
Pilon version 1.18 Sun Jun 12 01:01:53 2016 -0400
Genome: ./quiver62x_fast_v1.2/quiver62x_fast.contigs.fasta
Fixing bases, gaps, local
Input genome size: 494080101
Scanning BAMs
./quiver62x_fast_v1.2/LIPE_2_quiver62x_fast_v1.2.sorted.bam: 173934910 reads, 0 filtered, 172630192 mapped, 1648074
50 proper, 7574136 stray, FR 17% 334+/-80, RF 83% 4111+/-665, max 6107
./quiver62x_fast_v1.2/vu_all.sorted.bam: -2104535320 reads, 0 filtered, -2124364718 mapped, 2059806692 proper, 1065
60870 stray, FR 100% 199+/-34, max 2074
Processing tig00000191:1-56341
...
Processing tig00000504:1-45083
tig00000372:1-35214 log:
jumps ./quiver62x_fast_v1.2/LIPE_2_quiver62x_fast_v1.2.sorted.bam: coverage 3
frags ./quiver62x_fast_v1.2/vu_all.sorted.bam: coverage 12
Total Reads: 11317, Coverage: 15, minDepth: 5
Confirmed 20135 of 35214 bases (57.18%)
Corrected 35 snps; 0 ambiguous bases; 50 small insertions totaling 52 bases; 3 small deletions totaling 3 bases
# Attempting to fix local continuity breaks
# fix break: tig00000372:1562-2190 0 -0 +0 NoSolution
# fix break: tig00000372:2424-8487 0 -0 +0 NoSolution
...
Processing tig00000477:1-31833
tig00000504:1-45083 log:
jumps ./quiver62x_fast_v1.2/LIPE_2_quiver62x_fast_v1.2.sorted.bam: coverage 6
frags ./quiver62x_fast_v1.2/vu_all.sorted.bam: coverage 247
Total Reads: 172519, Coverage: 253, minDepth: 25
Confirmed 38095 of 45083 bases (84.50%)
Corrected 0 snps; 0 ambiguous bases; 3 small insertions totaling 3 bases; 1 small deletions totaling 1 bases
Large collapsed region: tig00000504:9597-32534 size 22938
# Attempting to fix local continuity breaks
# fix break: tig00000504:1351-1456 0 -0 +0 NoSolution
# fix break: tig00000504:1688-1913 0 -0 +0 NoSolution
...
# fix break: tig00000504:41923-42429 0 -0 +0 NoSolution
# fix break: tig00000504:42774-42985 0 -0 +0 NoSolution
Finished processing tig00000504:1-45083
Processing tig00001152:1-52833
tig00000965:1-79476 log:
jumps ./quiver62x_fast_v1.2/LIPE_2_quiver62x_fast_v1.2.sorted.bam: coverage 8
frags ./quiver62x_fast_v1.2/vu_all.sorted.bam: coverage 36
Total Reads: 62921, Coverage: 44, minDepth: 5
Confirmed 68624 of 79476 bases (86.35%)
Corrected 12 snps; 0 ambiguous bases; 10 small insertions totaling 15 bases; 0 small deletions totaling 0 bases
# Attempting to fix local continuity breaks
# fix break: tig00000965:4165-5831 0 -0 +0 NoSolution
# fix break: tig00000965:6280-6555 0 -0 +0 NoSolution TandemRepeat 191
...
```

**Comment by StefanoLonardi** [comment]
I checked other unitigs, and the coverage for those are much higher, for example
```
Processing tig00000854:1-190561
tig00001130:1-187448 log:
jumps ./quiver62x_fast_v1.2/LIPE_2_quiver62x_fast_v1.2.sorted.bam: coverage 22
frags ./quiver62x_fast_v1.2/vu_all.sorted.bam: coverage 363
Total Reads: 1145196, Coverage: 385, minDepth: 39
Confirmed 181576 of 187448 bases (96.87%)
Corrected 0 snps; 0 ambiguous bases; 19 small insertions totaling 20 bases; 0 small deletions totaling 0 bases
Large collapsed region: tig00001130:16354-29625 size 13272
Large collapsed region: tig00001130:65891-88825 size 22935
Large collapsed region: tig00001130:91830-110075 size 18246
# Attempting to fix local continuity breaks
```
so maybe everything is ok.

**Comment by w1bw** [comment]
Hi Stefano,

This is almost certainly because you have more than 2^31 reads, and I'm probably using a 32-bit integer to count the reads. Are these Illumina reads you are using to polish the CANU+quiver assembly?

If you are mostly concerned about cleaning up the base accuracy in the resulting assembly, I suggest running with "--fix bases" to avoid the local reassemblies (which take a lot of time and memory, especially with that much coverage!), and you can probably do just as well with half that much coverage.

**Comment by StefanoLonardi** [comment]
Yes, I am using Illumina reads to polish the CANU/quiver assembly. I will stop Pilon and run it again with the `--fix bases`. Thanks for the suggestion and for maintaining such a useful tool. Maybe in the next release you can fix the count limit of 2^31 reads.

**Comment by rmarquezp** [comment]
Hello, I'm in a similar situation while attempting to polish a 4.2Gb assembly. Pilon reports a negative number of reads for  one of my .bam files, which has 2,496,172,718  reads. The number of propperly paired reads does get counted correctly, and nother file with 1,087,897,567 reads does as well. I am attempting to polish a ~4.2Gb genome. Judging by the thread above I'd think this is just due to the large number of reads, but according to Pilon version history the problem with negative numbers of reads seems to have been fixed (I'm using version 1.23), so I'm not sure. Any idea of what could be causing this?

I mapped reads using bwa 0.7.17 and samtools 0.1.19 as follows:

`bwa mem -t 120 P_RNA_scaffold.fasta ~/reads/Illumina/PE/AllPE_R1.fastq.gz ~/reads/Illumina/PE/AllPE_R2.fastq.gz | samtools view - -Sb -@ 24 > PE_polish.bam
samtools sort -@ 120 -m 4gb PE_polish.bam PE_polish_sorted
samtools index PE_polish_sorted.bam`

And ran Pilon 1.23:

`java -Xmx2100G -jar ~/software/pilon-1.23.jar --genome P_RNA_scaffold.1.fasta --fix all --changes --frags ../PE_polish_sorted.bam --jumps ../MP_polish_sorted.bam --threads 56 --diploid --verbose --output PilonCleanup1`


The first few lines of the output are below:

`Pilon version 1.23 Mon Nov 26 16:04:05 2018 -0500
Genome: P_RNA_scaffold.1.fasta
Fixing snps, indels, gaps, local
Input genome size: 212260971
Scanning BAMs
../PE_polish_sorted.bam: 1087897567 reads, 0 filtered, 958121739 mapped, 309214556 proper, 92549442 stray, FR 100% 134+/-106, max 825
../MP_polish_sorted.bam: -1798794578 reads, 0 filtered, -2126931858 mapped, 519244786 proper, 244869104 stray, FR 100% 146+/-389, max 11956
Processing P_RNA_scaffold_2034:1-76459
Processing P_RNA_scaffold_1238:1-29182
Processing P_RNA_scaffold_542:1-843
Processing P_RNA_scaffold_5108:1-9195
Processing P_RNA_scaffold_5990:1-44018
Processing P_RNA_scaffold_6123:1-19701
.....
`

Thanks!
